### PR TITLE
Add UDS support

### PR DESCRIPTION
**Pull request by violetagg**
Covers TCP transport

Related to #403

### CODE REVIEW

File: src/main/java/reactor/netty/resources/PooledConnectionProvider.java [hunk 1]
```diff
@@ -634,7 +635,7 @@ void registerClose(PooledRef<PooledConnection> pooledRef, InstrumentedPool<Poole
 
 		PoolKey(SocketAddress holder, int pipelineKey) {
 			this.holder = holder;
-			this.fqdn = holder instanceof InetSocketAddress ? holder.toString() : "null";
+			this.fqdn = holder instanceof InetSocketAddress || holder instanceof DomainSocketAddress ? holder.toString() : "null";
```
Comment (ericbottard) [inline src/main/java/reactor/netty/resources/PooledConnectionProvider.java:638]: Should this condition be inverted and `toString()` invoked for all non-null cases instead?

File: src/main/java/reactor/netty/udp/UdpClientConfig.java [hunk 2]
```diff
@@ -74,15 +74,26 @@ public final InternetProtocolFamily family() {
...
+
+	@Override
+	protected ChannelFactory<? extends Channel> connectionFactory(EventLoopGroup elg, boolean isDomainSocket) {
+		if (isDomainSocket) {
+			throw new UnsupportedOperationException();
+		}
+		DatagramChannel channel;
 		if (isPreferNative()) {
-			channelFactory = () -> loopResources().onChannel(DatagramChannel.class, elg);
+			channel = loopResources().onChannel(DatagramChannel.class, elg);
```
Comment (ericbottard) [inline src/main/java/reactor/netty/udp/UdpClientConfig.java:91]: This changes the semantics of when objects are created (once vs at factory invocation). Is this intended?

File: src/test/java/reactor/netty/tcp/TcpClientTests.java [hunk 3]
```diff
@@ -1029,4 +1030,35 @@ private void checkReference(WeakReference<ByteBuf> ref) throws Exception {
...
+		         .connectNow();
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void testTcpClientWithDomainSocketsWithPort() {
+		TcpClient.create()
+		         .remoteAddress(() -> new DomainSocketAddress("/tmp/test.sock"))
+		         .port(1234)
+		         .connectNow();
+	}
```
Comment (ericbottard) [inline src/test/java/reactor/netty/tcp/TcpClientTests.java:1063]: Should there be new tests for things that *are* supported?

File: src/test/java/reactor/netty/tcp/TcpServerTests.java [hunk 4]
```diff
@@ -934,6 +936,77 @@ public void testGracefulShutdown() throws Exception {
...
+	public void testTcpServerWithDomainSocketsWithPort() {
+		TcpServer.create()
+		         .bindAddress(() -> new DomainSocketAddress("/tmp/test.sock"))
+		         .port(1234)
+		         .bindNow();
+	}
+
+	@Test
+	public void testTcpServerWithDomainSockets() throws Exception {
+		if (!LoopResources.hasNativeSupport()) {
```
Comment (ericbottard) [inline src/test/java/reactor/netty/tcp/TcpServerTests.java:972]: Seems like a could candidate for JUnit `assumeTrue()`

File: src/test/java/reactor/netty/tcp/TcpServerTests.java [hunk 5]
```diff
@@ -934,6 +936,77 @@ public void testGracefulShutdown() throws Exception {
...
+		    .sendString(Flux.just("1", "2", "3"))
+		    .then()
+		    .subscribe();
+
+		CountDownLatch latch = new CountDownLatch(1);
+		conn.inbound()
+		    .receive()
+		    .asString()
+		    .doOnNext(s -> {
+		        if (s.endsWith("3")) {
```
Comment (ericbottard) [inline src/test/java/reactor/netty/tcp/TcpServerTests.java:998]: Why `endsWith` and not `equals`?

File: src/test/java/reactor/netty/tcp/TcpClientTests.java (hunk 3 above)
Comment (ericbottard) [reply src/test/java/reactor/netty/tcp/TcpClientTests.java:1063]: I see the server tests covers the client part, nvm

File: src/test/java/reactor/netty/udp/UdpServerTests.java [hunk 6]
```diff
@@ -269,4 +270,27 @@ public void portBindingException() {
...
+		         .bindNow();
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void testUdpServerWithDomainSocketsWithPort() {
+		UdpServer.create()
+		         .bindAddress(() -> new DomainSocketAddress("/tmp/test.sock"))
+		         .port(1234)
+		         .bindNow();
+	}
```
Comment (ericbottard) [inline src/test/java/reactor/netty/udp/UdpServerTests.java:295]: Should there be a test for a valid server+client interaction, similar to the TCP case?

Comment (violetagg) [reply src/test/java/reactor/netty/udp/UdpServerTests.java:295]: no support for UDP protocol

File: src/test/java/reactor/netty/tcp/TcpServerTests.java (hunk 5 above)
Comment (violetagg) [reply src/test/java/reactor/netty/tcp/TcpServerTests.java:998]: There is no guarantee how the messages will be received on the server side, this depends on the buffers on network level and the flushes that will be performed.


For example here with this test
The client sends
```
16:18:28.996 [reactor-tcp-kqueue-2] DEBUG reactor.netty.tcp.TcpClient - [id: 0xe6610674, L:null - R:/tmp/test.sock] WRITE: 1B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 31                                              |1               |
+--------+-------------------------------------------------+----------------+
16:18:28.998 [reactor-tcp-kqueue-2] DEBUG reactor.netty.tcp.TcpClient - [id: 0xe6610674, L:null - R:/tmp/test.sock] WRITE: 1B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 32                                              |2               |
+--------+-------------------------------------------------+----------------+
16:18:28.999 [reactor-tcp-kqueue-2] DEBUG reactor.netty.tcp.TcpClient - [id: 0xe6610674, L:null - R:/tmp/test.sock] WRITE: 1B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 33                                              |3               |
+--------+-------------------------------------------------+----------------+
```

on the server this is the incoming and outgoing data:

```
16:18:29.002 [reactor-tcp-kqueue-3] DEBUG reactor.netty.tcp.TcpServer - [id: 0x2b5371dd] READ: 3B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 31 32 33                                        |123             |
+--------+-------------------------------------------------+----------------+
16:18:29.004 [reactor-tcp-kqueue-3] DEBUG reactor.netty.tcp.TcpServer - [id: 0x2b5371dd] WRITE: 3B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 31 32 33                                        |123             |
+--------+-------------------------------------------------+----------------+
```

File: src/main/java/reactor/netty/resources/PooledConnectionProvider.java (hunk 1 above)
Comment (violetagg) [reply src/main/java/reactor/netty/resources/PooledConnectionProvider.java]: let's call `toString` in all cases

File: src/main/java/reactor/netty/udp/UdpClientConfig.java (hunk 2 above)
Comment (violetagg) [reply src/main/java/reactor/netty/udp/UdpClientConfig.java]: thanks for spotting this -> changed

File: src/test/java/reactor/netty/tcp/TcpServerTests.java (hunk 4 above)
Comment (violetagg) [reply src/test/java/reactor/netty/tcp/TcpServerTests.java]: changed

**Comment by violetagg** [comment]
> Tried to review as much as I could, given my limited understanding of the architecture for now.
> 
> One thing that strikes me though is the numerous use of instanceof decisions. Maybe this warrants introduction of a strategy pattern?

Sure. Please propose a change.

File: src/main/java/reactor/netty/resources/DefaultLoopEpoll.java [hunk 7]
```diff
@@ -50,6 +54,12 @@
 		if (channelClass.equals(DatagramChannel.class)) {
 			return (CHANNEL) new EpollDatagramChannel();
 		}
+		if (channelClass.equals(DomainSocketChannel.class)) {
+			return (CHANNEL) new EpollDomainSocketChannel();
+		}
+		if (channelClass.equals(ServerDomainSocketChannel.class)) {
+			return (CHANNEL) new EpollServerDomainSocketChannel();
```
Comment (simonbasle) [inline src/main/java/reactor/netty/resources/DefaultLoopEpoll.java:61]: I think that is orthogonal to this PR, but I'm wondering about the throwing in both `getChannel` and `channelType`. AFAICS when producing a `ChannelFactory` in `xxxConfig#connectionFactory`, the lambda representing the factory calls `channelType`, which can throw. The result is then wrapped by a call to `getChannel`, which can also throw. But that exception comes late in the cycle, _after_ having called the configuration methods, right ? Or is the exception surfacing right away when building the client / binding the server?

I was wondering if a null-object pattern could be used instead to signal an illegal combination of channel type and transport earlier. The `channelType` would produce a special class and the factory  could check it eagerly, in order to throw from the configuration phase?

of course this is all moot if the factory MUST be lazy (ie, the channel type can be changed BETWEEN a call to `xxxConfig#connectionFactory` and the final build/bind step. This is also less of an issue if the throwing `ChannelFactory` can never make it past the build/bind step, ie the resulting client or server could never start.

File: src/main/java/reactor/netty/transport/TransportConfig.java [hunk 8]
```diff
@@ -236,12 +236,33 @@ protected TransportConfig(TransportConfig parent) {
...
+	 *   <li>{@link io.netty.channel.unix.DomainSocketChannel}</li>
+	 *   <li>{@link io.netty.channel.unix.ServerDomainSocketChannel}</li>
+	 *   <li>{@link io.netty.channel.socket.DatagramChannel}</li>
+	 * </ul>
+	 *
+	 * @param isDomainSocket true if {@link io.netty.channel.unix.DomainSocketChannel} or
+	 * {@link io.netty.channel.unix.ServerDomainSocketChannel} is needed, false otherwise
+	 * @return the channel type this configuration is associated with
+	 */
+	protected abstract Class<? extends Channel> channelType(boolean isDomainSocket);
```
Comment (simonbasle) [inline src/main/java/reactor/netty/transport/TransportConfig.java:253]: can we expect new type of sockets to be supported at some point? (see if an enum would make sense). It seems there's a lot of dimensions now: client vs server + transport protocol + isNative + isDomainSocket. Is there a simplification possible due to mutually exclusive dimensions? (eg. can we collapse the two booleans into a single enum?)

File: src/main/java/reactor/netty/resources/DefaultLoopEpoll.java (hunk 7 above)
Comment (violetagg) [reply src/main/java/reactor/netty/resources/DefaultLoopEpoll.java:61]: By default `LoopResources#onChannel` method invokes the default implementation used by Reactor Netty.
It throws exception to signal that the default implementation does not know what to do with other class types. Of course the user can provide his/her own implementation that will be able to handle other class types.

File: src/main/java/reactor/netty/transport/TransportConfig.java (hunk 8 above)
Comment (violetagg) [reply src/main/java/reactor/netty/transport/TransportConfig.java:253]: I'll think about that and address on a later stage.

**Pull request merged**